torchtune/models/convert_weights.py

```diff
@@ -38,8 +38,10 @@
     "model.layers.{}.mlp.gate_proj.weight": "layers.{}.mlp.w1.weight",
     "model.layers.{}.mlp.up_proj.weight": "layers.{}.mlp.w3.weight",
     "model.layers.{}.mlp.down_proj.weight": "layers.{}.mlp.w2.weight",
-    "model.layers.{}.input_layernorm.weight": "layers.{}.sa_norm.scale",
-    "model.layers.{}.post_attention_layernorm.weight": "layers.{}.mlp_norm.scale",
+    "model.layers.{}.input_layernorm.weight": "layers.{}.sa_norm.scale",  # mlp_norm.scale -> looks like a previous bug here  # noqa
+    "model.layers.{}.post_attention_layernorm.weight": "layers.{}.sa_scale.scale",
```

Comment:
@ebsmothers what do you think about this line?
It felt like the correct mapping between hugging face's implementation and torchtune implementation would be:
`"model.layers.{}.post_attention_layernorm.weight": "layers.{}.sa_scale.scale"`
instead of this:
`"model.layers.{}.post_attention_layernorm.weight": "layers.{}.mlp_norm.scale",`

I was comparing this from hugging face: https://github.com/huggingface/transformers/blob/3a10c6192b12b328685dce89004f9f9b9c042d30/src/transformers/models/gemma2/modular_gemma2.py#L54

But when looking at gemma first implementation here it seems that there is a mismatch between the two versions of hugging face:
https://github.com/huggingface/transformers/blob/3a10c6192b12b328685dce89004f9f9b9c042d30/src/transformers/models/gemma/modular_gemma.py#L791

It looks like a bug from HF side which has been duplicated in torchtune, what do you think ? I will try to ask on hugging face repo and see how it goes.

Official Google's implementation is here: https://github.com/google/gemma_pytorch/blob/80881c2e6e797ef1913a4a705d4b40394791cc58/gemma/model.py#L446

Reply:
Actually I think the mismatch comes from the original repo with different naming for the same normalization:
https://github.com/google/gemma_pytorch/blob/80881c2e6e797ef1913a4a705d4b40394791cc58/gemma/model.py#L383

So there should not be any bug at all, but something has to be done in torchtune to handle both cases.
Should I create a specific weight converter dictionnary for gemma 2 in torchtune ?

Reply:
Thanks @Optimox for the investigation on this. I see what you're saying: the Hugging Face `post_attention_layernorm` is applied on the residual version in Gemma but on the raw self-attention outputs in Gemma 2, but uses the same name in both cases.

In that case, I think it's best to create a separate Gemma 2 model type. It'll use an almost-duplicate version of Gemma's key mapping, but since the same key gets mapped to different values in Gemma vs Gemma 2 I don't think there's any other way. Maybe just make sure to add a comment in the key mapping to explain that this is the only difference between the two.